### ISSUE TITLE
Variable Block Scope

### ISSUE THREAD

**Issue by nickl-** [beanshell/beanshell#555]
BeanShell follows variable block scope for untyped variables as per JAVA.

This was introduced in ver 2.0b6 already, previously untyped variables were exposed globally. 

Variables are only visible within the scope of the block (which includes child blocks) in which they were declared. If your script relied on the old behaviour the easiest fix would be to simply declare the untyped variable, within the scope it is used, before it gets assigned.

```java

myvar = null;  // declare untyped variable by assigning it any value

if (true) {
    myvar = true; // original untyped variable assignment within a block
} 

assigned = myvar; // variable now in scope

```

This is contradicted by the [documentation](http://beanshell.org/manual/syntax.html#Basic_Scoping_of_Variables) which needs to be updated, as identified by @tamminenih #554

**Comment by morris821028** [comment]
We don't have the similar keyword `global` in python [Python function global variables](https://stackoverflow.com/questions/10588317/python-function-global-variables). 

```
unset("myvar");

void func() {
    if (myvar == void)
        myvar = 1;  // limited accessibility in this condition.
    myvar = myvar+1;  // limited accessibility in function.
}

func();
assigned = myvar;
```

We need to ensure the variable scope before calling function. I think it is a complex test case for nested script function. It could be difficult to export the variable to global scope.

**Comment by nickl-** [comment]
### BeanShell NameSpace

Each block scope has its own name space you can access the name space via: `this.namespace`

```java
bsh % test() {
    return this.namespace;
}
--> void
bsh % test();
--> $1 = NameSpace: test (bsh.NameSpace@7d6f77cc) (method)  :bsh.NameSpace
```

You can access the caller via: `this.caller` 

```java
bsh % test() {
    return this.caller.namespace;
}
--> void
bsh % test();
--> $2 = NameSpace: global (bsh.NameSpace@2d363fb3) :bsh.NameSpace
```

BeanShell also has a `global` you can access the global name space via: `global.namespace`

```java
bsh % test() {
    return global.namespace;
}
--> void
bsh % test();
--> $3 = NameSpace: global (bsh.NameSpace@2d363fb3) :bsh.NameSpace
```

### Changing NameSpace

You can change the current name space anywhere with the command `setNameSpace()` and your scope will change to that name space.

```java
bsh % test() {
     setNameSpace( global.namespace );
     local = 5;
     return this.namespace;
}
--> void
bsh % local;
--> void
bsh % test();
--> $1 = NameSpace: global (bsh.NameSpace@2d363fb3) :bsh.NameSpace
bsh % local;
--> $2 = 5I :int
```

For more information on [setNameSpace](http://beanshell.org/manual/bshmanual.html#setNameSpace()) see the documentation.

### NameSpace variables

You can simply assign a new variable as a property directly on the name space. To add a variable to the `global` name space scope:

```java
bsh % unset("local");
--> void
bsh % test() {
    local = 6543;
    global.local = local;
}
--> void
bsh % local;
--> void
bsh % test();
--> $6 = 6543I :int
bsh % local;
--> $7 = 6543I :int
```

The same works for example using `this.caller` to assign a variable in the caller's scope.

```java
bsh % unset("local");
--> void
bsh % test() {
    local = 6543;
    this.caller.local = local;
}
--> void
bsh % local;
--> void
bsh % test();
--> $8 = 6543I :int
bsh % local;
--> $9 = 6543I :int
```

More information on [Scope Modifiers](http://beanshell.org/manual/bshmanual.html#Scope_Modifiers) in the documentation.

### Interpreter globals

There is another way to add a global variable by setting it on the interpreter. You can access the interpreter anywhere via `this.interpreter` regardless of the scope. To set a global variable use `this.interpreter.set(name, value)`:

```java
bsh % test() {
    another_local = 1234;
    this.interpreter.set("another_local", another_local);
}
--> void
bsh % another_local;
--> void
bsh % test();
--> void
bsh % another_local;
--> $5 = 1234I :int
```
This also works from within Java to [accessing variables in the interpreter](http://beanshell.org/manual/bshmanual.html#Calling_BeanShell_From_Your_Application) wit `set()` and `get()`.

**NOTE: The documentation is not up to date but there is still useful information there. Please keep in mind that things may have changed and might not work as advertised.**

We will get the documentation updated this is what these [discussions](https://github.com/beanshell/beanshell/issues?q=is%3Aissue+is%3Aopen+label%3Adiscussion) are for.

**Comment by opeongo** [comment]
Is this why the following used to work, but now does not?

```java
if (true) {
  a = 1;
}
print(a);
```

As mentioned this works just fine in Bsh-2.0b4 (and is documented as how it works).  In the current development version the print statement will fail with 'Evaluation Error: Undefined argument'

Is there a flag or switch that can turn the old behaviour back on?

**Comment by nickl-** [comment]
@opeongo your statement is correct, in order for your code to work you need a reference to variable `a` outside the block.

**Comment by opeongo** [comment]
This change breaks a lot of code that I am supporting.  This behaviour is still documented as a feature implemented on purpose in the BeanShell documentation.  I am not sure why changing this to default to strict java mode is a good thing.  Breaking compatibility with an advertised feature seems like a questionable decision.

In my case it means that if I want to use Bsh 3.0 I will have to make a fork and revert this.

Is this a change that would be better if it only went in with strictjava mode?

**Comment by nmaqsudov** [comment]
> This change breaks a lot of code that I am supporting.

@opeongo , the same is happened to me. 
The least change we can do to make the code

```
if (true) {
  a = 1;
}
print(a);
```

is working as previously is adding a `super` qualifier

```
if (true) {
  super.a = 1;
}
print(a);   // now works as 2.0b4
```

**Comment by opeongo** [comment]
I had a look in to the code to see how to revert this or perhaps make it conditional.  It looks to me like this was an inadvertent change that occurred in 2016 when property access was added.  A new method `setVariableOrProperty` was added to the `NameSpace` class and the `BlockNameSpace` sub-class did not get updated as well.  If you want to revert this change then that is where to start.  There is some dead code there that really needs to be cleaned up, but the technical debt is a bit frightening.

It would be possible to revert this, but I also feel like it is time to move on.  Allowing this loose definition of variables was questionable in the first place (so non-Java and contrary to all Java documentation) and it is best to pay the price now and get over it.

I suggest to close this issue.

**Comment by nickl-** [comment]
> It would be possible to revert this, but I also feel like it is time to move on. Allowing this loose definition of variables was questionable in the first place (so non-Java and contrary to all Java documentation) and it is best to pay the price now and get over it.

Agreed, it was broken before and is now fixed. 

> For people like me who have customers with large codes bases larded with this problem it simply isn't an option to make this change. I thought I could deal with the new behaviour, but it involves way too much churn. If you are interested you can get this change from my copy of this repository.

Is it going to be that difficult to add a reference outside the block? I am sure we can write a script that can search through the source code and patch the fix.

Lets take step back and address this question from another angle... is there any other language in existence that breaks variable scope in this way?

The real question: is it a feature or a bug?

**Comment by opeongo** [comment]
Sigh...  this is so unfortunate.

The original design decision in Bsh-0.0.1 to export block scope variables in to the parent was a bug.  No other language is like this.  It saves a few keystrokes but it is un-Java and goes against reasonable expectations (principle of least surprise).

However, it was a decision that was made.  It is documented.  Some existing BeanShell code depends on it.  To change it now will break some existing scripts and make upgrading more difficult than just a drag-n-drop.

Let's look at other communities and how they handle bad design decisions.  One of the Linus Torvalds commandments is ["We don't break userspace"](https://lkml.org/lkml/2012/12/23/75) and he has been pretty emphatic about this.  Once userspace depends on a feature or an API then in general it cannot be changed without expending a lot of effort.  A big part of the linux success story is built on backwards compatibility.  The same goes with Java which is very conservative.  Most Java 1.1 compiled classes will still run under the latest version.  Consider the opposite situation, like moving from Python 2.7 to Python 3 which took years because of multiple language incompatibilities.  Or the move from Java 8 to Java 9+, where many users are sticking with Java 8 due to library compatibility issues.

If I was developing a language from scratch I would not have allowed this feature.  But I feel like it is here and we need to acknowledge it and accommodate it in some way.  Just removing the old behaviour and saying ['live with it'](https://lore.kernel.org/all/Pine.LNX.4.64.0512290923420.14098@g5.osdl.org/) will break trust and needlessly make life harder for people who we want in our community.  

I feel like there needs to be some relatively simple configuration option that can be used to turn the old behaviour back on.  Something that can be described in the release notes and in the FAQ and be toggled on by a user with just a few key strokes (maybe with admonishments to fix the scripts and get with the new program).  Because if people upgrade and their scripts break, it would be nice to have some simple advice for how to overcome the problem quickly and carry on.  If we want Bsh-3.0 to be popular then we need to reduce the amount of friction that people have to come on in.

There is a huge palette of colours for this bike shed.  Perhaps with a runtime flag to enable the old behaviour (-Dbsh.variable.blockscope=true) with the default value to be set to false.  Or maybe a method on the interpreter instance to turn it on or off.   Or maybe an environment variable.  Or a setting in an rc file.  Maybe something that can be put in to a script file so that it can be applied selectively within each script Namespace context.  The implementing code wouldn't be difficult, it literally comes down to a conditional test with one additional line in the else clause.

**Comment by nickl-** [comment]
> However, it was a decision that was made. It is documented. Some existing BeanShell code depends on it. To change it now will break some existing scripts and make upgrading more difficult than just a drag-n-drop.

It actually changed a long time ago, was part of the merged forks code when block scope was finally implemented. I will argue that it was never "designed" that way per se, but simply a happens stance resulting from the missing scope implementation which was advocated as a "cool feature" and documented during the original promotion so as to avoid having to address bug reports when more pressing things were outstanding. 

The documentation is so far removed from reality that it could today be construed as mere fiction rather than actual fact. Another "cool feature" that happened by chance or necessity during the evolution of Beanshell which we can reminisce on but it holds no weight.

> Let's look at other communities and how they handle bad design decisions.

We need to compare apples with apples. Beanshell migrated over 3 different repositories (left in different states) and had a hard fork which was finally merged (including communities), it even had a complete change of license at a stage. At no point was there ever stable release (non beta) which could be considered as a legacy to maintain and it is impossible to know what the code out there actually ran against because there were several releases in the wild. 

I agree with the notion of keeping backward compatibility but only once you decided that you actually have something to maintain, and definitively implementing scope would've been a part of it...which is why it was a priority to those who came before us.
     
> needlessly make life harder for people who we want in our community.

I am not saying there is no empathy, for starters it will need to be documented as well as documentation for upgrade gotchas and I even suggested last week we script something that will fix old code. That sounds like support to me and not a "live with it" attitude.

> We should make the new behaviour the default, but I feel like there needs to be some relatively simple configuration option that can be used to turn the old behaviour back on. .... The implementing code wouldn't be difficult, it literally comes down to a conditional test with one additional line in the else clause.

=) 

I understand you are nostalgic but I don't think you have realized what the impact is of this "trivial to implement" configuration is to even Beanshell code when scope variables start to bleed out into their parents. Would the stack even cope and how or where do you presume we stop it? On class level, on package, or do we just turn everything global? Beanshell today is way more complex than back in the day and still has lots outstanding, like any other language it relies on scope to safe guard others from pollution. We are also java still, and rely on garbage collection to clean up lost references so they therefor necessarily need to get lost...hence the need for scope.

Just of the top of my head I can only imagine the can of worms this will open but the reality, and finding the cause when the yet unimagined problems occur will not be fun to put it mildly.

Bottom line:
This isn't something new in Beanshell 3.0 it has actually been there for many years and the decisions were already made, thankfully, long before we got here. The only reason there is still code like that in the wild is because it is running against very old beta libraries. Reintroducing this bug is going to cause more problems than it solves and we should instead focus the limited resources on something more productive. 

I am not unsympathetic to anyone's plight and we could work on a simple patch script that could just scan the code and insert references where required. It doesn't even have to parse the code simple regex captures would suffice in building a tree. 

Then I would prefer we first find out if anyone is actually going to use it. If there is code running against such old libraries which never needed to be upgraded, what is to say they will now choose to upgrade? If they do then all that tells me is that there is a need to change the old into something new so it is already part of the objective. 

It needs to be justified to build the parser, so much more so to build bloat into the interpreter which the majority of users will never need. The pandora's box you are trying to open should also beware, I guarantee that this is not the only incompatibility there will be. Does that mean we break everything again in the spirit of backward compatibility? Of course it does!!

It is not cool, it is not nice to have, it is just broken.

**Comment by jimjag** [comment]
Closing during a backlog triage pass: the maintainers affirmed in the thread that the current Java-like scoping is correct and intentional. Reopen if this is still relevant.
